File: backend/app/schemas/watchlist.py
"""Watchlist Pydantic schemas - 用户自选股请求与响应模型。

提供：
- WatchlistAddRequest: 加入自选请求（仅 instrument_id + source，user_id 由上下文注入）
- WatchlistItemResponse: 单条自选响应
- WatchlistListResponse: 自选列表响应
- WatchlistMonitorStatusItem: 自选股+监控状态聚合响应（单条）
- WatchlistMonitorStatusResponse: 自选股+监控状态聚合响应（列表）

安全约束：
- user_id 不出现在请求体中（由认证上下文注入）
- instrument_id 为 UUID
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class WatchlistAddRequest(BaseModel):
    """加入自选请求。

    user_id 不接受客户端传入（由认证上下文注入），仅传 instrument_id 与来源。
    """

    instrument_id: UUID = Field(..., description="股票 ID（UUID）")
    source: str = Field(
        default="manual",
        description="加入来源（manual/selection_plan/monitor）",
    )


class WatchlistItemResponse(BaseModel):
    """单条自选响应。"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID = Field(..., description="自选记录 ID")
    user_id: UUID = Field(..., description="用户 ID")
    instrument_id: UUID = Field(..., description="股票 ID")
    source: str = Field(..., description="加入来源")
    active: bool = Field(..., description="是否活跃")
    created_at: datetime = Field(..., description="加入时间")
    removed_at: datetime | None = Field(None, description="移除时间")


class WatchlistListResponse(BaseModel):
    """自选列表响应。"""

    items: list[WatchlistItemResponse] = Field(default_factory=list, description="自选列表")
    total: int = Field(..., description="总记录数")


class WatchlistMonitorStatusItem(BaseModel):
    """自选股+监控状态聚合响应（单条）。"""

    watchlist_item_id: UUID = Field(..., description="自选记录 ID")
    instrument_id: UUID = Field(..., description="股票 ID")
    symbol: str = Field(..., description="股票代码")
    name: str = Field(..., description="股票名称")
    market: str = Field(..., description="市场（SH/SZ/BJ）")
    watchlist_created_at: datetime = Field(..., description="加入自选时间")
    monitor_status: str = Field(
        ..., description="监控状态枚举：WAITING_FIRST_RUN/SUCCEEDED/FAILED/STALE/MARKET_CLOSED"
    )
    evaluation_status: str | None = Field(
        None, description="评估状态（SUCCEEDED/FAILED/PENDING）"
    )
    error_code: str | None = Field(
        None, description="评估错误码（无错误时为 null）"
    )
    source_bar_time: str | None = Field(
        None, description="监控状态对应的 bar 时间"
    )
    metrics: dict[str, Any] | None = Field(
        None, description="监控状态 metrics（MonitorState payload，无状态时为 null）"
    )
    updated_at: datetime | None = Field(
        None, description="监控状态更新时间"
    )


class WatchlistMonitorStatusResponse(BaseModel):
    """自选股+监控状态聚合响应（列表）。"""

    items: list[WatchlistMonitorStatusItem] = Field(
        default_factory=list, description="自选股+监控状态列表"
    )


if __name__ == "__main__":
    # 自测入口：验证 schema 字段定义
    print(f"WatchlistAddRequest fields={list(WatchlistAddRequest.model_fields.keys())}")
    print(f"WatchlistItemResponse fields={list(WatchlistItemResponse.model_fields.keys())}")
    print(f"WatchlistListResponse fields={list(WatchlistListResponse.model_fields.keys())}")
    print(f"WatchlistMonitorStatusItem fields={list(WatchlistMonitorStatusItem.model_fields.keys())}")
    print(f"WatchlistMonitorStatusResponse fields={list(WatchlistMonitorStatusResponse.model_fields.keys())}")
    # 验证 user_id 不在请求体中（安全约束）
    assert "user_id" not in WatchlistAddRequest.model_fields, "user_id 不应在请求体中"
    print("OK")

File: frontend/src/api/watchlist.ts
// Watchlist 领域 API owner
//
// [S3-B] 由 endpoints.ts 迁出。endpoints.ts 仅保留兼容 barrel 重新导出，
// 既有 caller 的 import path 与运行时行为完全不变。
//
// 依赖方向：client ← market ← watchlist ← endpoints(barrel) / useWatchlistApi。
// WatchlistMonitorStatusItem 复用 market.MarketSession（单一 owner，不复制）。

import { apiClient } from './client'
import type { MarketSession } from './market'

// ============================================================
// Watchlist 领域类型
// ============================================================

/** 自选股项 */
export interface WatchlistItem {
  id: string
  user_id: string
  instrument_id: string
  source: string
  active: boolean
  created_at: string
  removed_at: string | null
}

/** 自选股列表响应 */
export interface WatchlistListResponse {
  items: WatchlistItem[]
  total: number
}

/** 自选股 metadata-only summary 项（与 backend WatchlistSummaryItem 对齐）
 *  P0 后续数据边界：仅 instrument 元数据 + 加入时间，禁止行情/策略指标 */
export interface WatchlistSummaryItem {
  watchlist_item_id: string
  instrument_id: string
  symbol: string
  name: string
  market: string
  source: string
  created_at: string
}

/** 自选股 metadata-only summary 列表响应（GET /v1/watchlist 专用） */
export interface WatchlistSummaryResponse {
  items: WatchlistSummaryItem[]
  total: number
}

/** 自选股+监控状态聚合项（与 backend/app/schemas/watchlist.py WatchlistMonitorStatusItem 对齐） */
export interface WatchlistMonitorStatusItem {
  watchlist_item_id: string
  instrument_id: string
  symbol: string
  name: string
  market: string
  watchlist_created_at: string
  monitor_status: MarketSession | 'WAITING_FIRST_RUN' | 'SUCCEEDED' | 'FAILED' | 'STALE'
  market_session: MarketSession
  calculation_status: 'SUCCEEDED' | 'FAILED' | 'STALE' | 'WAITING_FIRST_RUN'
  freshness_seconds: number | null
  last_bar_time: string | null
  evaluation_status: string | null
  retry_count: number | null
  error_code: string | null
  source_bar_time: string | null
  metrics: Record<string, unknown> | null
  latest_event?: {
    event_type: string
    event_time: string
    boundary: number | null
  } | null
  updated_at: string | null
}

/** 自选股+监控状态聚合响应 */
export interface WatchlistMonitorStatusResponse {
  items: WatchlistMonitorStatusItem[]
}

/** 加入自选请求 */
export interface WatchlistAddRequest {
  instrument_id: string
  source?: string
}

// ============================================================
// ===== Watchlist 端点 =====
// ============================================================

/** 查询当前用户的自选列表（metadata-only summary，仅 active=true） */
export async function getWatchlist(): Promise<WatchlistSummaryResponse> {
  const { data } = await apiClient.get<WatchlistSummaryResponse>('/v1/watchlist')
  return data
}

/** 加入自选（instrument_id，user_id 由认证上下文注入） */
export async function addToWatchlist(payload: WatchlistAddRequest): Promise<WatchlistItem> {
  const { data } = await apiClient.post<WatchlistItem>('/v1/watchlist', payload)
  return data
}

/** 移除自选（软删除：active=false + removed_at） */
export async function removeFromWatchlist(instrumentId: string): Promise<void> {
  await apiClient.delete(`/v1/watchlist/${instrumentId}`)
}

/** 查询自选股+监控状态聚合数据 */
export async function getWatchlistMonitorStatus(): Promise<WatchlistMonitorStatusResponse> {
  const { data } = await apiClient.get<WatchlistMonitorStatusResponse>('/v1/watchlist/monitor-status')
  return data
}
